ambari-server/src/main/java/org/apache/ambari/server/controller/internal/DefaultProviderModule.java

```diff
@@ -132,6 +132,10 @@ protected ResourceProvider createResourceProvider(Resource.Type type) {
         return new ArtifactResourceProvider(managementController);
       case RemoteCluster:
         return new RemoteClusterResourceProvider();
+      case Host:
+        return new HostResourceProvider(managementController);
+      case HostSummary:
+        return new HostInfoSummaryResourceProvider(managementController);
```

Comment:
This is interesting - I would have expected you to use the same ResourceProvider here (`HostResourceProvider`) and simply change the renderer in the `HostResourceDefinition`. Is there a reason a whole new ResourceProvider is needed instead of just a new renderer?

Reply:
At begining, I thought to reuse HostResourceProvider, but I found they are totoally different resource provider: 
(1) HostResourceProvider focuses on create, update and delete Host, but HostInfoSummaryResourceProvider is a read-only provider
(2) HostResourceProvider is only on single host, but the latter will aggregate info from all hosts in cluster or accross clusters
So I think it is more clear to separete these two kinds of host resource providers.

Reply:
HostResourceProvider does do multiple hosts though:

```
    Set<HostResponse> responses = getResources(new Command<Set<HostResponse>>() {
      @Override
      public Set<HostResponse> invoke() throws AmbariException {
        return getHosts(requests);
      }
    });
```

It iterates over the `HostResponse` instances, creating resources for each. So, wouldn't you only need to supply a renderer in this case to post-process the results?

Reply:
The reason is performance. My implementation is to sum os_type of all the hosts on DB level with single sql call. If we do post-process for all the hosts after creating resource for each, in large number of hosts scenario, it will take too long to get resources for all hosts.

Reply:
I also thought we will be adding a new renderer like we do for Alerts. I understand that this is more performant but it is a tradeoff between performance & maintenance. If the host summary API is not going to be used very often then we should consider just going with a renderer. Can you check with the UI team on how often the host summary api will be called in the UI.

Reply:
I will check UI about the use case. In considerationof maintenance, I think the new implementation is more clear and clean. The host resouce provider is onlyb for single host crud, the hosts summary resource provider is a read provideer and works like a aggregation service and provide overall hosts summary in/across clusters. Now I only implement os_type sum, but it is very easy to add more inforamtion in the future.

Reply:
The host resource provider does provide information for multiple hosts. If we're introducing a new resource provider, then we shouldn't need a renderer since the default renderer can be used (but it should be a new endpoint). So I think we need to either:
- Switch to using a renderer with the current resource provider
or
- Create a new endpoint and new resource provider with a default renderer for the host summary